storage/interceptable_repository.go

```diff
@@ -232,6 +232,35 @@ func (ir *queryScopedInterceptableRepository) Delete(ctx context.Context, object
 	return nil
 }
 
+func (ir *queryScopedInterceptableRepository) ForceDelete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
```

Comment:
Spare the code duplication by extracting the common logic and refactoring the `Delete`, `DeleteReturning` and `ForceDelete` methods like so:

```go
func (ir *queryScopedInterceptableRepository) DeleteReturning(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) (types.ObjectList, error) {
	var resultList types.ObjectList
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		var err error
		if resultList, err = ir.repositoryInTransaction.DeleteReturning(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	if err := ir.delete(ctx, objectType, criteria, deleteObjectFunc); err != nil {
		return nil, err
	}

	return resultList, nil
}

func (ir *queryScopedInterceptableRepository) Delete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		if err := ir.repositoryInTransaction.Delete(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	return ir.delete(ctx, objectType, criteria, deleteObjectFunc)
}

func (ir *queryScopedInterceptableRepository) ForceDelete(ctx context.Context, objectType types.ObjectType, criteria ...query.Criterion) error {
	deleteObjectFunc := func(ctx context.Context, _ Repository, _ types.ObjectList, deletionCriteria ...query.Criterion) error {
		if err := ir.repositoryInTransaction.ForceDelete(ctx, objectType, deletionCriteria...); err != nil {
			return err
		}
		return nil
	}

	return ir.delete(ctx, objectType, criteria, deleteObjectFunc)
}

func (ir *queryScopedInterceptableRepository) delete(ctx context.Context, objectType types.ObjectType, criteria []query.Criterion, deleteObjectFunc InterceptDeleteOnTxFunc) error {
	if deleteOnTxFunc, found := ir.deleteOnTxFuncs[objectType]; found {
		objects, err := ir.List(ctx, objectType, criteria...)
		if err != nil {
			return err
		}
		delete(ir.deleteOnTxFuncs, objectType)
		if err := deleteOnTxFunc(deleteObjectFunc)(ctx, ir, objects, criteria...); err != nil {
			ir.deleteOnTxFuncs[objectType] = deleteOnTxFunc
			return err
		}
		ir.deleteOnTxFuncs[objectType] = deleteOnTxFunc

	} else {
		if err := deleteObjectFunc(ctx, nil, nil, criteria...); err != nil {
			return err
		}
	}

	return nil
}
```